### ISSUE TITLE
Define the concept of an "opaque string"

### ISSUE THREAD

**Issue by mr-c** [common-workflow-language/common-workflow-language#914]
https://en.wikipedia.org/wiki/Opaque_data_type

**Comment by mr-c** [comment]
My understanding is that we meant

> A nonsensical value that is swapped out with a real value later in the evaluation process. Probably some sort of easy to recognize random string

@tetron can provide more context

**Comment by tetron** [comment]
Yes, that's basically it.  The important thing is that workflow/tool expressions shouldn't try to parse it.

**Comment by tetron** [comment]
The purpose being to allow the runner to swap in a value late in the evaluation process, after the workflow/tool expressions have had a chance to see it.

**Comment by kinow** [comment]
Today-I-Learned about opaque data types, and opaque string.

I had a look at the Wikipedia article, and also at the text where we use “opaque”, but I feel like it would be better to replace it by another term instead of clarifying it or adding a link, so we don't interrupt the reader.

**Comment by kinow** [comment]
I spent some time reading more about opaque strings, and it appears to be used in specs, some programming languages, and I found this Microsoft 365 [doc](https://docs.microsoft.com/en-us/microsoft-365/cloud-storage-partner-program/online/scenarios/business), this AWS [doc](https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/cloudhsmv2/model/DescribeBackupsResult.html), and this NASA [doc](https://wiki.earthdata.nasa.gov/display/DOIsforEOSDIS/ESDSWG+Recommendations) on DOI that use the term as well.

Maybe we should leave “opaque” and “opaque string”, but clarify it somewhere else in the document? What others think? If not, maybe a replacement term could be placeholder string, always followed by a note that implementees must not rely on this value…

**Comment by mr-c** [comment]
We could insert a glossary after https://www.commonwl.org/v1.2/CommandLineTool.html#Terminology and define the terms there; they can be linked back from where they are used
